packages/react-server/src/ReactFizzServer.js

```diff
@@ -3239,10 +3310,31 @@ function queueCompletedSegment(
   }
 }
 
+// Merges the internal state of Hoistables from a source to a target. afterwards
+// both source and target will share the same unified state. This technique is used
+// so we can resolve tasks in any order and always accumulate up to the root Hoistable.
+// This function uses the `this` argument to allow for slightly optimized calling from
+// a forEach over a Set.
+function mergeHoistables(this: Hoistables, source: Hoistables) {
+  if (enableFloat) {
+    const target = this;
+    hoistHoistables(target.state, source.state);
+    source.fallbacks.forEach(h => {
+      target.fallbacks.add(h);
+    });
+    // we assign the parent state and fallbacks to the child state so if a child task completes
+    // it will hoist and merge with the parent Hoistables
+    source.state = target.state;
+    source.fallbacks = target.fallbacks;
+  }
+}
+
 function finishedTask(
   request: Request,
   boundary: Root | SuspenseBoundary,
   segment: null | Segment,
+  hoistables: Hoistables,
+  parentHoistables: null | Hoistables,
```

Comment:
The parentHoistables is really reasoning on a distance here that I'm not really comfortable reasoning about. Especially in the resuming path where the nearest boundary isn't necessarily even reified. A resumed boundary is just skipped.